chain/chain/src/update_shard.rs

```diff
@@ -232,8 +252,8 @@ fn apply_old_chunk(
 
     let storage_config = RuntimeStorageConfig {
         state_root: *prev_chunk_extra.state_root(),
-        use_flat_storage: true,
-        source: crate::types::StorageDataSource::Db,
+        use_flat_storage: shard_context.storage_data_source == StorageDataSource::Db,
```

Comment:
I suggest to keep `use_flat_storage: true` here and move this logic as part of `apply_transactions` implementation: 
```
StorageDataSource::DbTrieOnly => {
   let mut trie = self.get_trie_for_shard(
                shard_id,
                prev_block_hash,
                storage_config.state_root,
                false,
            );
  trie.dont_charge_gas_for_trie_node_access();
  trie;
}
```
just to avoid duplication and changing it back after we have state witness